CI: Lint starlark files with `buildifier`

jdbaldry opened the pull request:
**What is this feature?**

Adds a pipeline step to lint Starlark files in CI.

**Why do we need this feature?**

`buildifier` ensures a consistent format for all Starlark files which makes the readers experience more consistent.
It also provides additional linting such as checking the presence of module and function docstrings, or requesting the removal of loaded functions that are unused. I believe this will also help new readers understand the code faster.

**Who is this feature for?**

Primarily release engineering engineers.

**Which issue(s) does this PR fix?**:

None

**Special notes for your reviewer**:

jdbaldry committed:
Add verify-starlark build action that returns an error for starlark files with lint

Relies on `buildifier` tool.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Add verify_starlark_step to PR pipeline

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Manually fetch buildifier in curl_image until a new build_image is created

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Format with buildifier

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Remove all unused variables retaining one unused function

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Use snake_case for variable

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Replace deprecated dictionary concatenation with .update() method

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Start adding docstrings for all modules and functions

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Prefer os.WriteFile as ioutil.WriteFile has been deprecated since go 1.16

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Attempt to document the behavior of the init_enterprise_step

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document test_backend pipeline

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document enterprise_downstream_step

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document the pipeline utility function

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document publish_images_step

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document publish_images_steps

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document enterprise2_pipelines function

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Add tags table for Starlark files.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document test_frontend

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document windows function

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Add docstrings to verifystarlark functions

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

kminehart commented on pkg/build/cmd/verifystarlark.go:
```diff
@@ -0,0 +1,84 @@
+			cmd.Dir = workspace
+			output, err := cmd.CombinedOutput()
+			if err != nil {
+				switch err.Error() {
+				case "exit status 1": // syntax errors in input
+					errs = append(errs, fmt.Errorf("unexpected syntax error from command %q: %s", cmd, string(output)))
+				case "exit status 2": // usage errors: invoked incorrectly
+					errs = append(errs, fmt.Errorf("unexpected usage error from command %q: %s", cmd, string(output)))
+				case "exit status 3": // unexpected runtime errors: file I/O problems or internal bugs
+					errs = append(errs, fmt.Errorf("unexpected runtime error from command %q: %s", cmd, string(output)))
+				case "exit status 4": // check mode failed (reformat is needed)
+					errs = append(errs, errors.New(string(output)))
+				default:
+					errs = append(errs, fmt.Errorf("unexpected error from command %q: %v", cmd, err))
+				}
```
My only criticism is that it feels a little shaky to be comparing the error string.

I would rather see the `*exec.ExitError` and `errors.As` here and using the `ExitCode` int.

https://pkg.go.dev/os/exec#ExitError

jdbaldry committed:
Refactor error handling to be more clear and document complex behavior

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Split errors into execution errors and verification errors

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document all other library functions

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Add local variables to TAGS

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Add blank line between all Args and Returns sections

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Merge remote-tracking branch 'origin/main' into jdb/2022-11-lint-starlark

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Fix new linting errors

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Lint new Starlark files

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Merge remote-tracking branch 'origin/main' into jdb/2022-11-lint-starlark

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Correct buildifier binary mv

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Document the need to set nofile ulimit to at least 2048

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Update build-container to include buildifier

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Merge remote-tracking branch 'origin/main' into jdb/2022-11-lint-starlark

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Ensure buildifier binary is executable

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

zerok commented on Makefile:
```diff
@@ -233,8 +233,12 @@ drone: $(DRONE)
+	etags --lang none --regex="/def \(\w+\)[^:]+:/\1/" --regex="/\s*\(\w+\) =/\1/" $^ -o $@
+
 format-drone:
-	black --include '\.star$$' -S scripts/drone/ .drone.star
```
Wow, wasn't aware that black also somehow worked with starlark 🤯

jdbaldry replied:
Haha, I wouldn't have thought of it either. I believe it was added because of Starlark's syntactic similarity to Python.

zerok commented on pkg/build/cmd/verifystarlark.go:
```diff
@@ -0,0 +1,153 @@
+	workspace := c.Args().Get(0)
+	verificationErrs, executionErr := verifyStarlark(c.Context, workspace, buildifierLintCommand)
+	if executionErr != nil {
+		return cli.Exit(executionErr.Error(), 1)
```
What would be the outcome if you'd just return the executionErr instead of doing cli.Exit in here?

jdbaldry replied:
Good thought, I've tested locally and it behaves the same. I was just following the pattern of other `cli.Exit` usage throughout the code because I'm not familiar with the library.

zerok commented on pkg/build/cmd/verifystarlark.go:
```diff
@@ -0,0 +1,153 @@
+	// If any execution of the `buildifier` command fails, it is reported as executionErr and
+	// any verification of subsequent files is skipped.
+	if err := filepath.WalkDir(workspace, func(path string, d fs.DirEntry, err error) error {
+		// Skip verification of the file or files within the directory if there is an error
```
Please also check inside the WalkFunc if the context has been cancelled and return the context's error if so.

jdbaldry replied:
I'm also not very familiar with using the `context` package. How should one correctly check the channel returned by `context.Done()` in a WalkFunc such as this?

zerok replied:
You normally do it somehow like this:

```go
select {
  case <- ctx.Done():
    return ctx.Err()
  default:
}
```

In the ctx.Done case the err will be most likely the reason why the context has been cancelled, so you could either take this one as executionErr or use some other error to just get out of the walk callback 😄

jdbaldry replied:
Just added a test case to check the context and realized that we already handle context cancellation automatically when I pass the context to each command execution in `cmd := exec.CommandContext(ctx, command, args...)`. Do you still think it is worthwhile to check the context before running any command?

zerok replied:
I don't think that's necessary as any longrunning step would notice it anyway and abort anyway 🙂

zerok commented on pkg/build/cmd/verifystarlark.go:
```diff
@@ -0,0 +1,153 @@
+		return nil
+	}); err != nil {
+		// Should never happen. See comment by the invocation of WalkDir for more information.
+		panic(fmt.Sprintf("unexpected error from filepath.WalkDir: %v", err))
```
Couldn't you treat the `executionErr` as return value for the Walk here? Then you could leave that panic to the caller 🙂

jdbaldry replied:
That's an interesting point. I always struggle to mentally model the WalkFunc error handling so it might take me a little while to review this suggestion and check how things behave.

zerok replied:
Basically any error that is not filepath.SkipDir will just abort the whole Walk-process and get returned.

jdbaldry replied:
Ah! I think I had mistakenly believed that all errors from the `WalkFunc` are filtered through the next call in addition to the `WalkDir` func errors that are filtered through the `WalkFunc`

jdbaldry replied:
Done in fbe033ae25c

zerok reviewed (approved):
This looks awesome, thank you 😄 Just some tiny nitpicks but nothing that needs to be done right now 🟢

jdbaldry committed:
Fix valid content test

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Simply return execution error

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Only check files rather than fixing them

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Use updated build-container with executable buildifier

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Test that context cancellation stops execution

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Simplify error handling

Return execution errors that short circuit WalkDir rather than
separately tracking that error.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Merge remote-tracking branch 'origin/main' into jdb/2022-11-lint-starlark

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Remove fetching of buildifier binary now that it is in the build-container

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Use build image in verify-starlark step

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Use semver tag

The image is the same but uses a semver tag to make it clearer that
this is a forward upgrade from the old version.

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Merge remote-tracking branch 'origin/main' into jdb/2022-11-lint-starlark

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>

jdbaldry committed:
Use node 18 image with buildifier

Signed-off-by: Jack Baldry <jack.baldry@grafana.com>
